### PR TITLE
kubeadm: Remove multiple API server endpoints support upon join

### PR DESCRIPTION
**Pull request by rosti**
**What this PR does / why we need it**:

In the past the discovery configuration expected, that we can support multiple API server endpoints. In practice, we always end up with a single API server endpoint, because, even in HA setups, we use a load balancer scheme for API servers.
Therefore, to reduce complexity and improve readability of the config, the multiple API server endpoints support is removed from the bootstrap token discovery join method and configuration.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
refs kubernetes/kubeadm#911

**Special notes for your reviewer**:

Depends on:
- [X] #67763 

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/kind api-change
/kind cleanup
/assign @luxas
/assign @timothysc
/assign @fabriziopandini

**Release note**:

```release-note
kubeadm: Multiple API server endpoints support upon join is removed as it is now redundant.
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/join.go [hunk 1]
```diff
@@ -245,6 +249,26 @@ func AddJoinConfigFlags(flagSet *flag.FlagSet, cfg *kubeadmapiv1beta1.JoinConfig
 	)
 }
 
+// AddJoinBootstrapTokenDiscoveryFlags adds bootstrap token specific discovery flags to the specified flagset
+func AddJoinBootstrapTokenDiscoveryFlags(flagSet *flag.FlagSet, btd *kubeadmapiv1beta1.BootstrapTokenDiscovery) {
+	flagSet.StringVar(
+		&btd.Token, "discovery-token", "",
+		"A token used to validate cluster information fetched from the api server.")
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/join.go]: maybe take the opportunity to `api` -> `API` here?

File: cmd/kubeadm/app/cmd/join.go [hunk 2]
```diff
@@ -245,6 +249,26 @@ func AddJoinConfigFlags(flagSet *flag.FlagSet, cfg *kubeadmapiv1beta1.JoinConfig
...
+	flagSet.BoolVar(
+		&btd.UnsafeSkipCAVerification, "discovery-token-unsafe-skip-ca-verification", false,
+		"For token-based discovery, allow joining without --discovery-token-ca-cert-hash pinning.")
+}
+
+// AddJoinFileDiscoveryFlags adds file discovery flags to the specified flagset
+func AddJoinFileDiscoveryFlags(flagSet *flag.FlagSet, fd *kubeadmapiv1beta1.FileDiscovery) {
+	flagSet.StringVar(
+		&fd.KubeConfigPath, "discovery-file", "",
+		"A file or url from which to load cluster information.")
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/join.go]: `url` -> `URL`

**Review by neolit123 — COMMENTED** [review]
thanks, added a couple of minor comments.

File: cmd/kubeadm/app/cmd/join.go (hunk 1 above)
Comment (rosti) [reply cmd/kubeadm/app/cmd/join.go]: Fixed in #67763

File: cmd/kubeadm/app/cmd/join.go (hunk 2 above)
Comment (rosti) [reply cmd/kubeadm/app/cmd/join.go]: Fixed in #67763

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 3]
```diff
@@ -280,50 +280,72 @@ type JoinConfiguration struct {
...
-	DiscoveryToken string
-	// DiscoveryTokenAPIServers is a set of IPs to API servers from which info
-	// will be fetched. Currently we only pay attention to one API server but
-	// hope to support >1 in the future.
-	DiscoveryTokenAPIServers []string
-	// DiscoveryTimeout modifies the discovery timeout
-	DiscoveryTimeout *metav1.Duration
+
+	// Discovery specifies the options for the kubelet to use during the TLS Bootstrap process
+	Discovery Discovery
```
[the remark this answers is not in the document]
Comment (rosti) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: @neolit123 it appears, that up until now, the empty line is used to distinguish between different groups of fields in a single structure. Like the ones, who are dependent in one-another.

With the redesign of config, most of the related fields were moved into substructs and therefore single fields, unrelated to others have become more common. Hence most of the field structures have empty lines between them.

**Comment by rosti** [comment]
Waiting on #67763 to merge.

/hold

**Comment by rosti** [comment]
This is now ready for final review.

/hold cancel

**Review by neolit123 — APPROVED** [review]
/lgtm
thanks for the update.

File: cmd/kubeadm/app/cmd/join.go [hunk 4]
```diff
@@ -177,7 +177,7 @@ func NewCmdJoin(out io.Writer) *cobra.Command {
 				cfg.Discovery.File = fd
 			} else {
 				cfg.Discovery.BootstrapToken = btd
-				cfg.Discovery.BootstrapToken.APIServerEndpoints = args
+				cfg.Discovery.BootstrapToken.APIServerEndpoint = args[0]
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/join.go]: What about adding a check that args has one element?

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
@rosti many thanks for this PR!
only one minor nit and then ready to lgtm

**Comment by rosti** [comment]
@fabriziopandini updated with your feedback, also added another check in the conversion code to avoid possible panic.

**Review by fabriziopandini — APPROVED** [review]
/approve
IMO this PR is now ready to merge. Please rebase to get the test passing

/assign @timothysc

File: cmd/kubeadm/app/discovery/token/token.go [hunk 5]
```diff
@@ -181,42 +181,40 @@ func buildSecureBootstrapKubeConfig(endpoint string, caCert []byte, clustername
...
 	select {
 	case <-time.After(discoveryTimeout):
-		close(stopChan)
+		once.Do(func() {
+			close(cancelChan)
+		})
 		err := fmt.Errorf("abort connecting to API servers after timeout of %v", discoveryTimeout)
 		fmt.Printf("[discovery] %v\n", err)
-		wg.Wait()
+		<-finishedChan
```
Comment (timothysc) [inline cmd/kubeadm/app/discovery/token/token.go]: I think the wait/waitgroup is a lot cleaner here imo.  

The double use of the channel for both the case statement and within the case is an anti-pattern.

**Review by timothysc — CHANGES_REQUESTED** [review]

**Review by fabriziopandini — APPROVED** [review]
@rosti  Thanks for this PR!
/approve
/lgtm

**Pull request merged**